lib/mayaUsd/ufe/UsdUndoVisibleCommand.cpp

```diff
@@ -39,19 +42,26 @@ UsdUndoVisibleCommand::Ptr UsdUndoVisibleCommand::create(const UsdPrim& prim, bo
     if (!prim) {
         return nullptr;
     }
-    return std::make_shared<UsdUndoVisibleCommand>(prim, vis);
-}
 
-void UsdUndoVisibleCommand::execute()
-{
-    UsdGeomImageable primImageable(_prim);
+    auto layer = getEditRouterLayer(PXR_NS::TfToken("visibility"), prim);
+
+    UsdGeomImageable primImageable(prim);
+
+    EditTargetGuard guard(prim, layer);
 
     std::string errMsg;
     if (!MayaUsd::ufe::isAttributeEditAllowed(primImageable.GetVisibilityAttr(), &errMsg)) {
```

Comment:
Perform the edit check on command creation, rather than execution.  Failure on execution should be reported through an exception, but here the check really belongs to command creation: we communicate to the caller that the operation is not possible by returning a nullptr.

Furthermore, the edit guard must be set before performing the check, otherwise the check will be done on the current target, then the edit guard changes that target.  In the case of an existing visibility edit in the session layer, this would mean that the check would be done on the current edit target layer, which typically is lower priority than the session layer, and the edit would fail.  If the check is done when the edit target guard is set, the check properly occurs on the session layer.  This accounts for many of the failures described in

https://github.com/Autodesk/maya-usd/issues/2549